velox/exec/TableWriter.cpp

```diff
@@ -260,43 +267,73 @@ void TableWriter::close() {
   }
 }
 
-void TableWriter::setConnectorOrWriterMemoryReclaimer(
-    memory::MemoryPool* pool) {
-  VELOX_CHECK_NOT_NULL(pool);
-  if (operatorCtx_->pool()->reclaimer() != nullptr) {
-    pool->setReclaimer(
-        TableWriter::MemoryReclaimer::create(operatorCtx_->driverCtx(), this));
+void TableWriter::setConnectorMemoryReclaimer() {
+  VELOX_CHECK_NOT_NULL(connectorPool_);
+  if (connectorPool_->parent()->reclaimer() != nullptr) {
+    connectorPool_->setReclaimer(TableWriter::ConnectorReclaimer::create(
+        operatorCtx_->driverCtx(), this, spillConfig_.has_value()));
   }
 }
 
-std::unique_ptr<memory::MemoryReclaimer> TableWriter::MemoryReclaimer::create(
+std::unique_ptr<memory::MemoryReclaimer>
+TableWriter::ConnectorReclaimer::create(
     DriverCtx* driverCtx,
-    Operator* op) {
+    Operator* op,
+    bool canReclaim) {
   return std::unique_ptr<memory::MemoryReclaimer>(
-      new TableWriter::MemoryReclaimer(
-          driverCtx->driver->shared_from_this(), op));
+      new TableWriter::ConnectorReclaimer(
+          driverCtx->driver->shared_from_this(), op, canReclaim));
 }
 
-bool TableWriter::MemoryReclaimer::reclaimableBytes(
+bool TableWriter::ConnectorReclaimer::reclaimableBytes(
     const memory::MemoryPool& pool,
     uint64_t& reclaimableBytes) const {
-  VELOX_CHECK(!pool.isLeaf());
   reclaimableBytes = 0;
-  return false;
-}
-
-uint64_t TableWriter::MemoryReclaimer::reclaim(
-    memory::MemoryPool* pool,
-    uint64_t /*unused*/,
-    memory::MemoryReclaimer::Stats& /*unused*/) {
-  VELOX_CHECK(!pool->isLeaf());
-  return 0;
+  if (!canReclaim_) {
+    return false;
+  }
+  std::shared_ptr<Driver> driver = ensureDriver();
+  if (FOLLY_UNLIKELY(driver == nullptr)) {
+    return false;
+  }
+  return memory::MemoryReclaimer::reclaimableBytes(pool, reclaimableBytes);
 }
 
-void TableWriter::MemoryReclaimer::abort(
+uint64_t TableWriter::ConnectorReclaimer::reclaim(
     memory::MemoryPool* pool,
-    const std::exception_ptr& /* error */) {
-  VELOX_CHECK(!pool->isLeaf());
+    uint64_t targetBytes,
+    memory::MemoryReclaimer::Stats& stats) {
+  if (!canReclaim_) {
+    return 0;
+  }
+  std::shared_ptr<Driver> driver = ensureDriver();
+  if (FOLLY_UNLIKELY(driver == nullptr)) {
+    return 0;
+  }
+  VELOX_CHECK(
+      !driver->state().isOnThread() || driver->state().isSuspended ||
+      driver->state().isTerminated);
+  VELOX_CHECK(driver->task()->pauseRequested());
+
+  auto* writer = dynamic_cast<TableWriter*>(op_);
+  if (writer->closed_) {
+    // TODO: reduce the log frequency if it is too verbose.
+    ++stats.numNonReclaimableAttempts;
+    LOG(WARNING) << "Can't reclaim from a closed writer connector pool: "
```

Comment:
Why do we need to log these warnings when we already have runtime stats that can tell us whether this is happening or not and how often?